flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableFsDataOutputStream.java

```diff
@@ -175,44 +172,39 @@ private static void safelyTruncateFile(
 		}
 	}
 
-	private static void ensureTruncateInitialized() throws FlinkRuntimeException {
-		if (HadoopUtils.isMinHadoopVersion(2, 7) && truncateHandle == null) {
-			Method truncateMethod;
+	private static Method findTrunacteMethod(FileSystem fs) throws FlinkRuntimeException {
+		Method truncateMethod = null;
+		if (HadoopUtils.isMinHadoopVersion(fs.getClass().getClassLoader(), 2, 7)) {
 			try {
-				truncateMethod = FileSystem.class.getMethod("truncate", Path.class, long.class);
-			}
-			catch (NoSuchMethodException e) {
+				truncateMethod = fs.getClass().getMethod("truncate", Path.class, long.class);
+			} catch (NoSuchMethodException e) {
 				throw new FlinkRuntimeException("Could not find a public truncate method on the Hadoop File System.");
 			}
-
-			if (!Modifier.isPublic(truncateMethod.getModifiers())) {
-				throw new FlinkRuntimeException("Could not find a public truncate method on the Hadoop File System.");
-			}
-
-			truncateHandle = truncateMethod;
 		}
+
+		return truncateMethod;
 	}
 
-	private static boolean truncate(final FileSystem hadoopFs, final Path file, final long length) throws IOException {
-		if (!HadoopUtils.isMinHadoopVersion(2, 7)) {
-			throw new IllegalStateException("Truncation is not available in hadoop version < 2.7 , You are on Hadoop " + VersionInfo.getVersion());
+	private static boolean truncate(
```

Comment:
change to non static method and use `truncateHandle` from the instance?

Reply:
> Do we need some new test coverage?

Yes, but I don't have a good idea for a non-complicated test setup. In theory, we would need to have a test matrix where we co-ship Flink with hadoop 2.6 and 2.8 and have plugins with hadoop 2.6 and 2.8.

Reply:
Yes, that's what I'd wanted to do but unfortunately, `HadoopRecoverableWriter` accesses the static `HadoopRecoverableFsDataOutputStream.HadoopFsCommitter` which in turn invokes the truncate.
We could factor out a common TruncateWrapper though.